guppylang/std/_internal/compiler/array.py

```diff
@@ -180,44 +186,158 @@ def compile(self, args: list[Wire]) -> list[Wire]:
             return [self.build_classical_array(args)]
 
 
+ARRAY_GETITEM_CLASSICAL: Final[GlobalConstId] = GlobalConstId.fresh()
+ARRAY_GETITEM_LINEAR: Final[GlobalConstId] = GlobalConstId.fresh()
+ARRAY_SETITEM_CLASSICAL: Final[GlobalConstId] = GlobalConstId.fresh()
+ARRAY_SETITEM_LINEAR: Final[GlobalConstId] = GlobalConstId.fresh()
+
+
 class ArrayGetitemCompiler(ArrayCompiler):
     """Compiler for the `array.__getitem__` function."""
 
-    def build_classical_getitem(self, array: Wire, idx: Wire) -> CallReturnWires:
-        """Lowers a call to `array.__getitem__` for classical arrays."""
+    def _construct_getitem_ty(self, bound: ht.TypeBound) -> ht.PolyFuncType:
+        """Constructs a polymorphic function type for `__getitem__`"""
+        # a(T, N), int -> T, a(T, N)
+        # Array element type parameter
+        elem_ty_param = ht.TypeTypeParam(bound)
+        # Array length parameter
+        length_param = ht.BoundedNatParam()
+        return ht.PolyFuncType(
+            params=[elem_ty_param, length_param],
+            body=ht.FunctionType(
+                input=[
+                    array_type(
+                        ht.Option(ht.Variable(0, bound)),
+                        ht.VariableArg(1, length_param),
+                    ),
+                    INT_T,
```

Comment:
Use `int_type().to_hugr()` (imported from `guppylang.tys.builtin`) here to make sure it's using the same bit-width as the rest of the compiler